### PR TITLE
Issue 137

### PR DESCRIPTION
**Pull request by jacobdotcosta**
Merged all hetzner related playbooks and docs from the ansible generic folder into the hetzner folder. Improved the documentation.

### CODE REVIEW

File: hetzner/README-cloud.md [hunk 1]
```diff
@@ -1,17 +1,82 @@
...
+
+All Ansible playbooks must be launched from the root of the project, at `k8s-infra`.
+
+## Init the hetzner context
+
 To getting started, you must get a Token for your API as described [here](https://docs.hetzner.cloud/#overview-getting-started).
 
 In order to create a vm and next access it, you must first import your ssh public key using this command
 ```bash
 hcloud ssh-key create --name USER_KEY_NAME --public-key-from-file ~/.ssh/id_rsa.pub
```
Comment (cmoulliard) [inline hetzner/README-cloud.md]: Is it still needed that we use this command as we use a shared ssh key when the vm is created ? If no, then disregard this section please

File: hetzner/README-cloud.md [hunk 2]
```diff
@@ -1,17 +1,82 @@
...
+| hetzner_context_name | x | | x | Context name for hcloud. |
+| vm_delete |  | False |  | Deletes the VM prior to creating it. |
+
+Each of the Ansible prompts can be replaced by defining it's value as an extra variable of the playbook.
+
+```bash
+$ ansible-playbook hetzner/ansible/hetzner-create-hcloud-server.yml -e '{ "vm_name": "my-name" }' -e '{ "salt_text": "<my salt>" }' -e '{ "password_text": "<my password>" }' -e '{ "hetzner_context_name": "<context name>" }' -e '{ "vm_delete": True }' -e '{ "pass_store_dir": "/home/<my home>/.my-password-store" }'
+```
+
+## Steps to create a k8s cluster
```
Comment (cmoulliard) [inline hetzner/README-cloud.md]: I will review this section after the merge and certainly move the commands to another README file or section of a README file

File: hetzner/ansible/hetzner-create-hcloud-server.yml [hunk 3]
```diff
@@ -62,7 +62,7 @@
       failed_when: "hcloud_delete_key_res.rc != 0 and ('SSH key not found' not in hcloud_delete_key_res.stderr)"
       when: "vm_delete is defined and vm_delete | bool"
 
-    - name: Create cloud server key
+    - name: "Create cloud server key - TODO: Use passwordstore"
```
Comment (cmoulliard) [inline hetzner/ansible/hetzner-create-hcloud-server.yml]: Use we use the `password store`, will the playbook git clone the project ?

File: hetzner/README-cloud.md (hunk 1 above)
Comment (jacobdotcosta) [reply hetzner/README-cloud.md]: The playbooks still use the local ssh key, the integration with pass is not implemented yet.

File: hetzner/ansible/hetzner-create-hcloud-server.yml (hunk 3 above)
Comment (jacobdotcosta) [reply hetzner/ansible/hetzner-create-hcloud-server.yml]: IMO we should leave the source of the `password store` project out of the scope of this playbook.

File: hetzner/README-cloud.md (hunk 1 above)
Comment (cmoulliard) [reply hetzner/README-cloud.md]: We have created a shared ssh key for the team -> snowdrop. So it is not needed to import the ssh key of the user. See our spring boot team info google doc -> Cloud Hetzner

File: hetzner/ansible/hetzner-create-hcloud-server.yml (hunk 3 above)
Comment (cmoulliard) [reply hetzner/ansible/hetzner-create-hcloud-server.yml]: Then, you must document that a pre-req will be to have the `password-store` project cloned locally

**Comment by cmoulliard** [comment]
Is this PR finalized @jacobdotcosta ?

**Comment by cmoulliard** [comment]
I did a test without success @jacobdotcosta 

```
~/Code/snowdrop/k8s-infra   issue-137  ansible-playbook hetzner/ansible/hetzner-create-server.yml \
>    -e vm_name=n01-k116 \
>    -e salt_text=$(head -c 500 /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 20 | head -n 1) \
>    -e password_text="xxxxxxxxx" \
>    -e hetzner_context_name=snowdrop \
>    -e override_public_key=~/.ssh/id_rsa_n01_k116.pub \
>    --tag "vm_delete"
[WARNING]: Skipping 'apiVersion' as this is not a valid group definition

[WARNING]: Skipping 'kind' as this is not a valid group definition

[WARNING]: Skipping 'clusters' as this is not a valid group definition

[WARNING]: Skipping 'contexts' as this is not a valid group definition

[WARNING]: Skipping 'current-context' as this is not a valid group definition

[WARNING]: Skipping 'users' as this is not a valid group definition

[WARNING]:  * Failed to parse /Users/dabou/Code/snowdrop/k8s-infra/ansible/inventory/id_openstack.rsa with ini plugin: /Users/dabou/Code/snowdrop/k8s-infra/ansible/inventory/id_openstack.rsa:1: Expected key=value host variable assignment, got: OPENSSH

[WARNING]: Unable to parse /Users/dabou/Code/snowdrop/k8s-infra/ansible/inventory/id_openstack.rsa as an inventory source

ERROR! the role 'hetzner/activate_context' was not found in /Users/dabou/Code/snowdrop/k8s-infra/hetzner/ansible/roles:/Users/dabou/roles:/Users/dabou/Code/snowdrop/k8s-infra/hetzner/ansible

The error appears to be in '/Users/dabou/Code/snowdrop/k8s-infra/hetzner/ansible/hetzner-create-server.yml': line 37, column 7, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

  roles:
    - role: hetzner/activate_context
      ^ here
```

**Comment by cmoulliard** [comment]
2 additional errors discovered
```
TASK [hetzner/create_server : Hash user password] ***********************************************************************************************************************************************************************************************************************
fatal: [127.0.0.1]: FAILED! => {"msg": "the field 'args' has an invalid value ({'hashed_user_pw': \"{{ password_text | password_hash('sha512', salt_text) }}\"}), and could not be converted to an dict.The error was: salt too large (sha512_crypt requires <= 16 chars)\n\nThe error appears to be in '/Users/dabou/Code/snowdrop/k8s-infra/ansible/roles/hetzner/create_server/tasks/main.yml': line 45, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Hash user password\n  ^ here\n"}

AND 

TASK [hetzner/create_server : Template cloudinit file] ******************************************************************************************************************************************************************************************************************
fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "AnsibleUndefinedVariable: 'str object' has no attribute 'ansible_user'"}
```

**Comment by cmoulliard** [comment]
2 additional errors discovered
```
TASK [hetzner/create_server : Hash user password] ***********************************************************************************************************************************************************************************************************************
fatal: [127.0.0.1]: FAILED! => {"msg": "the field 'args' has an invalid value ({'hashed_user_pw': \"{{ password_text | password_hash('sha512', salt_text) }}\"}), and could not be converted to an dict.The error was: salt too large (sha512_crypt requires <= 16 chars)\n\nThe error appears to be in '/Users/dabou/Code/snowdrop/k8s-infra/ansible/roles/hetzner/create_server/tasks/main.yml': line 45, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Hash user password\n  ^ here\n"}

AND 

TASK [hetzner/create_server : Template cloudinit file] ******************************************************************************************************************************************************************************************************************
fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "AnsibleUndefinedVariable: 'str object' has no attribute 'ansible_user'"}
```

**Comment by cmoulliard** [comment]
A new issue occurs during vm creation process @jacobdotcosta 

```
TASK [hetzner/create_server : Add server key to local authorized keys] ******************************************************************************************************************************
fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "Failed to lookup user centos: 'getpwnam(): name not found: centos'"}
```

**Comment by jacobdotcosta** [comment]
In the ansible ping of he hetzner, limit it to the host just created.

**Comment by cmoulliard** [comment]
I did a test locally and here are 2 errors reported when `pwd` dont exist like the `host_vars`
```
VM_NAME=h02-k115 ; PASSWORD_STORE_DIR=~/.password-store-snowdrop ; pass-rh rm snowdrop/hetzner/${VM_NAME} -rf  ; rm -f ~/.ssh/id_rsa_snowdrop_hetzner_${VM_NAME}* ; rm ansible/inventory/host_vars/${VM_NAME}
Error: snowdrop/hetzner/h02-k115 is not in the password store.
rm: ansible/inventory/host_vars/h02-k115: No such file or directory
```
@jacobdotcosta

**Comment by cmoulliard** [comment]
another issue on macos
```
TASK [passstore/ansible_inventory : Set pass facts (user, ssh_port)] *************************************************************************************************************************************************************************************************
task path: /Users/dabou/Code/snowdrop/k8s-infra/ansible/roles/passstore/ansible_inventory/tasks/generate_pass_structure.yml:5
fatal: [localhost]: FAILED! => {
    "msg": "An unhandled exception occurred while running the lookup plugin 'passwordstore'. Error was a <class 'ansible.errors.AnsibleError'>, original message: Command '['pass', 'insert', '-f', '-m', 'snowdrop/hetzner/h02-k115/os_user']' returned non-zero exit status 1."
}
```

**Comment by jacobdotcosta** [comment]
> ansible_inventory

The 1st error is not a problem as the manual cleanup scripts fail if the folder doesn't exist but that doesn't affect the process.

**Comment by cmoulliard** [comment]
LGTM. I replayed successfully the steps to create the VM. Can you fix the conflicts and merge it please ? @jacobdotcosta

**Comment by cmoulliard** [comment]
Next time, dont mix within this PR k8s cluster creation with VM creation please @jacobdotcosta 

Here are 2 issues discovered during k8s cluster creation or when resources where applied
```
TASK [k8s_config : Get CA] *******************************************************************************************************************************************************************************************************************************************
Wednesday 04 March 2020  17:05:21 +0100 (0:00:00.043)       0:03:46.076 *******
fatal: [h02-k115]: FAILED! => {"changed": true, "cmd": "kubectl --kubeconfig=$HOME/.kube/config -n kube-system get secret/ -o jsonpath='{.data.ca\\.crt}'\n", "delta": "0:00:00.044761", "end": "2020-03-04 17:05:22.554635", "msg": "non-zero return code", "rc": 1, "start": "2020-03-04 17:05:22.509874", "stderr": "error: stat /root/.kube/config: no such file or directory", "stderr_lines": ["error: stat /root/.kube/config: no such file or directory"], "stdout": "", "stdout_lines": []}
```

When resources are installed
```
TASK [ingress : Install Ingress] *************************************************************************************************************************************************************************************************************************************
Wednesday 04 March 2020  17:15:18 +0100 (0:00:02.826)       0:00:16.278 ******* 
fatal: [h02-k115]: FAILED! => {"changed": true, "cmd": "kubectl --kubeconfig=$HOME/.kube/config apply -f /tmp/ingress-configmap.yml\nkubectl --kubeconfig=$HOME/.kube/config apply -f /tmp/ingress-rbac.yml\nkubectl --kubeconfig=$HOME/.kube/config apply -f /tmp/ingress-svc.yml\nkubectl --kubeconfig=$HOME/.kube/config apply -f /tmp/ingress-dp.yml\n", "delta": "0:00:00.180942", "end": "2020-03-04 17:15:20.438076", "msg": "non-zero return code", "rc": 1, "start": "2020-03-04 17:15:20.257134", "stderr": "error: stat /root/.kube/config: no such file or directory\nerror: stat /root/.kube/config: no such file or directory\nerror: stat /root/.kube/config: no such file or directory\nerror: stat /root/.kube/config: no such file or directory", "stderr_lines": ["error: stat /root/.kube/config: no such file or directory", "error: stat /root/.kube/config: no such file or directory", "error: stat /root/.kube/config: no such file or directory", "error: stat /root/.kube/config: no such file or directory"], "stdout": "", "stdout_lines": []}
```
and
```
TASK [k8s_dashboard : Install Dashboard Version - v2.0.0-rc5] ********************************************************************************************************************************************************************************************************
Wednesday 04 March 2020  17:18:12 +0100 (0:00:12.056)       0:00:21.841 *******
fatal: [h02-k115]: FAILED! => {"changed": true, "cmd": "kubectl --kubeconfig=$HOME/.kube/config  apply -f  https://raw.githubusercontent.com/kubernetes/dashboard/v2.0.0-rc5/aio/deploy/recommended.yaml\n", "delta": "0:00:00.050934", "end": "2020-03-04 17:18:14.132831", "msg": "non-zero return code", "rc": 1, "start": "2020-03-04 17:18:14.081897", "stderr": "error: stat /root/.kube/config: no such file or directory", "stderr_lines": ["error: stat /root/.kube/config: no such file or directory"], "stdout": "", "stdout_lines": []}
```

**Comment by jacobdotcosta** [comment]
> Next time, dont mix within this PR k8s cluster creation with VM creation please @jacobdotcosta
Ok, I'll separate it next time.
> 
> Here are 2 issues discovered during k8s cluster creation or when resources where applied
> 
> ```
> TASK [k8s_config : Get CA] *******************************************************************************************************************************************************************************************************************************************
> Wednesday 04 March 2020  17:05:21 +0100 (0:00:00.043)       0:03:46.076 *******
> fatal: [h02-k115]: FAILED! => {"changed": true, "cmd": "kubectl --kubeconfig=$HOME/.kube/config -n kube-system get secret/ -o jsonpath='{.data.ca\\.crt}'\n", "delta": "0:00:00.044761", "end": "2020-03-04 17:05:22.554635", "msg": "non-zero return code", "rc": 1, "start": "2020-03-04 17:05:22.509874", "stderr": "error: stat /root/.kube/config: no such file or directory", "stderr_lines": ["error: stat /root/.kube/config: no such file or directory"], "stdout": "", "stdout_lines": []}
> ```
> 
> When resources are installed
> 
> ```
> TASK [ingress : Install Ingress] *************************************************************************************************************************************************************************************************************************************
> Wednesday 04 March 2020  17:15:18 +0100 (0:00:02.826)       0:00:16.278 ******* 
> fatal: [h02-k115]: FAILED! => {"changed": true, "cmd": "kubectl --kubeconfig=$HOME/.kube/config apply -f /tmp/ingress-configmap.yml\nkubectl --kubeconfig=$HOME/.kube/config apply -f /tmp/ingress-rbac.yml\nkubectl --kubeconfig=$HOME/.kube/config apply -f /tmp/ingress-svc.yml\nkubectl --kubeconfig=$HOME/.kube/config apply -f /tmp/ingress-dp.yml\n", "delta": "0:00:00.180942", "end": "2020-03-04 17:15:20.438076", "msg": "non-zero return code", "rc": 1, "start": "2020-03-04 17:15:20.257134", "stderr": "error: stat /root/.kube/config: no such file or directory\nerror: stat /root/.kube/config: no such file or directory\nerror: stat /root/.kube/config: no such file or directory\nerror: stat /root/.kube/config: no such file or directory", "stderr_lines": ["error: stat /root/.kube/config: no such file or directory", "error: stat /root/.kube/config: no such file or directory", "error: stat /root/.kube/config: no such file or directory", "error: stat /root/.kube/config: no such file or directory"], "stdout": "", "stdout_lines": []}
> ```
> 
> and
> 
> ```
> TASK [k8s_dashboard : Install Dashboard Version - v2.0.0-rc5] ********************************************************************************************************************************************************************************************************
> Wednesday 04 March 2020  17:18:12 +0100 (0:00:12.056)       0:00:21.841 *******
> fatal: [h02-k115]: FAILED! => {"changed": true, "cmd": "kubectl --kubeconfig=$HOME/.kube/config  apply -f  https://raw.githubusercontent.com/kubernetes/dashboard/v2.0.0-rc5/aio/deploy/recommended.yaml\n", "delta": "0:00:00.050934", "end": "2020-03-04 17:18:14.132831", "msg": "non-zero return code", "rc": 1, "start": "2020-03-04 17:18:14.081897", "stderr": "error: stat /root/.kube/config: no such file or directory", "stderr_lines": ["error: stat /root/.kube/config: no such file or directory"], "stdout": "", "stdout_lines": []}
> ```

What's the difference between this execution that failed and the previous one that succeeded?

**Comment by cmoulliard** [comment]
> What's the difference between this execution that failed and the previous one that succeeded?

They both failed as ansible uses root user where the kube conf file is gone `"error: stat /root/.kube/config: no such file or directory"`

**Comment by jacobdotcosta** [comment]
Regarding the changes to `ansible/roles/k8s_cluster/tasks/install.yml`, 

> > What's the difference between this execution that failed and the previous one that succeeded?
> 
> They both failed as ansible uses root user where the kube conf file is gone `"error: stat /root/.kube/config: no such file or directory"`

I don't understand the error, because the `Install Dashboard Version` task doesn't use become, it should be executed by the `ansible_user`, not root.

**Comment by cmoulliard** [comment]
We will delete it, otherwise we risk to have to rebase, merge things already merged top of master. Use a different issue and its issue-xxx. @jacobdotcosta

**Pull request merged**